sycl/tools/sycl-xocc/bin/sycl_xocc.py

```diff
@@ -0,0 +1,344 @@
+you won't have the neccesarry information as the command is generated before
+the file's have even started to be compiled (Perhaps there is that I am
+unaware of). So no kernel names and no idea how many SDx commands you'd need
+to generate per file (no idea how many kernels are in a file).
+
+This works around that by using an opt (kernelNameGen) pass that generates
+an intermediate file with the needed information that we eat up and can then
+loop over each kernel in a file. It's simple at the moment just kernel names,
+but could expand in the future to include optimization information for each
+kernel.
+"""
+
+from argparse import ArgumentParser
+import json
+from os import environ
+from pathlib import Path
+import shutil
+import subprocess
+import tempfile
+
+
+class TmpDirManager:
+    def __init__(self, tmpdir: Path, prefix: str, autodelete: bool):
+        self.prefix = prefix
+        self.tmpdir = tmpdir
+        self.autodelete = autodelete
+
+    def __enter__(self) -> Path:
+        self.dir = Path(tempfile.mkdtemp(
+            dir=self.tmpdir,
+            prefix=self.prefix
+            ))
+        return self.dir
+
+    def __exit__(self, type, value, traceback):
+        if (self.autodelete):
+            shutil.rmtree(self.dir)
+
+
+class DataflowLawyerManager:
+    def __init__(self, vitis_clang_bin_dir: Path):
+        self.vitis_clang_bin_dir = vitis_clang_bin_dir
+        self.mounted = False
+
+    def __enter__(self):
+        self.dflawyer = self.vitis_clang_bin_dir / "xilinx-dataflow-lawyer"
+        echo = Path(shutil.which("echo"))
+        if self.dflawyer.is_file():
+            print(f"Binding {self.dflawyer} to {echo}")
+            subprocess.run([
+                "sudo", "mount", "--bind", echo, self.dflawyer
+            ])
+            self.mounted = True
+
+    def __exit__(self, _, __, ___):
+        if self.mounted:
+            print(f"Unbinding {self.dflawyer}")
+            subprocess.run([
+                "sudo", "umount", self.dflawyer
+            ])
+
+
+class CompilationDriver:
+    def __init__(self, arguments):
+        self.outpath = Path(arguments.o)
+        self.tmp_root = arguments.tmp_root
+        self.clang_path = arguments.clang_path.resolve()
+        self.inputs = arguments.inputs
+        self.hls_flow = arguments.hls
+        self.vitis_bin_dir = arguments.vitis_bin_dir.resolve()
+        self.vitis_version = self.vitis_bin_dir.parent.name
+        self.outstem = self.outpath.stem
+        self.vitis_mode = arguments.target
+        # TODO: XILINX_PLATFORM should be passed by clang driver instead
+        self.xilinx_platform = environ['XILINX_PLATFORM']
+        if environ.get("XILINX_CLANG_39_BUILD_PATH") is not None:
+            vitis_clang_bin = Path(environ["XILINX_CLANG_39_BUILD_PATH"])
+        else:
+            vitis_clang_bin = (
+                self.vitis_bin_dir /
+                "/../lnx64/tools/clang-3.9-csynth/bin"
```

Comment:
Since you spent some time using `Path`, is there a more symbolic way to do `..`?

Reply:
indeed there is : `path.parent or path.parents[0]`. Change applied